File: python/langchain-lcel.py
import os
import sys
from dotenv import load_dotenv
from langchain_openai import ChatOpenAI, OpenAI
from langchain_core.runnables import Runnable
from langchain_core.prompts import ChatPromptTemplate, MessagesPlaceholder
from langchain_core.output_parsers.string import StrOutputParser


# Load environment values from the .env file (for the API key)
load_dotenv()

def main():
    # Check for the optional -p parameter and assign the value if present
    param_index = sys.argv.index('-p') if '-p' in sys.argv else -1
    user_message = """Which countries played the final in the FIFA world cup in 2010?
Who won the game? Which players scored?"""

    if param_index != -1 and len(sys.argv) > param_index + 1:
        user_message = sys.argv[param_index + 1]

    # Define the prompt template
    prompt_template = ChatPromptTemplate.from_messages([
        ("system", "You are a helpful assistant."),
        ("user", "{user_message}")])
    
    # Set up the OpenAI model
    model = ChatOpenAI(api_key=os.getenv("OPENAI_API_KEY"), model="gpt-4o")

    # Define a simple parser
    parser = StrOutputParser()

    # Create the chain
    chain = prompt_template | model | parser

    # Invoke the chain
    response = chain.invoke({"user_message": user_message})

    print("Response...")
    print(response)

if __name__ == "__main__":
    main()

File: js/langchain-lcel.js
import { config } from 'dotenv';
import { ChatOpenAI } from '@langchain/openai';
import { RunnableSequence } from '@langchain/core/runnables';
import { ChatPromptTemplate } from '@langchain/core/prompts';
import { StringOutputParser } from '@langchain/core/output_parsers';

// Load environment values from the .env file (for the API key)
config();

async function main() {
  const args = process.argv;
  const paramIndex = args.indexOf('-p');
  let userMessage = `Which countries played the final in the FIFA world cup in 2010?
Who won the game? Which players scored?`;

  if (paramIndex !== -1 && args.length > paramIndex + 1) {
    userMessage = args[paramIndex + 1];
  }

  // Define the prompt template
  const promptTemplate = ChatPromptTemplate.fromMessages([
    [ "system", "You are a helpful assistant." ],
    ["human", "{user_message}"]
  ]);



  // Set up the OpenAI model
  const model = new ChatOpenAI({
    apiKey: process.env.OPENAI_API_KEY,
    model: "gpt-4o"
  });

  // Define a simple parser
  const parser = new StringOutputParser();

  // Create the chain
  const chain = RunnableSequence.from([promptTemplate, model, parser]);

  // Invoke the chain
  const response = await chain.invoke({ user_message: userMessage });

  console.log("Response...");
  console.log(response);
};

main().catch(console.error);
